### ISSUE TITLE
Weird conflict with `values(...).on_conflict_do_nothing()`

### ISSUE THREAD

**Issue by icefoxen** [diesel-rs/diesel#1930]
### Versions

- **Rust:** 1.31
- **Diesel:** 1.3.3
- **Database:** Postgres 11
- **Operating System** Linux

### Feature Flags

- **diesel:** `["postgres", "extras"]`

## Problem Description

I have this code:

```rust
#[derive(Debug, Clone, PartialEq, Eq, Hash, Insertable)]
#[table_name = "crates"]
pub struct NewCrate {
    pub name: &str,
    pub version: &str,
    pub checksum: &str,
    pub yanked: bool,
}

fn insert_crate(conn: &mut diesel::pg::PgConnection) {
    use crate::schema::crates::dsl::*;
    let mut versions: Vec<models::NewCrate> = vec![];
    // Omitted: populate `versions` with data from elsewhere...
    
    let new_ids: Vec<i64> = diesel::insert_into(crates)
        .values(&versions)
        .on_conflict_do_nothing()
        .returning(id)
        .get_results(conn)
        .unwrap();
    // ...do other stuff.
}
```

This works fine.  However, when I change the query just by passing the vec by value instead of by reference:

```rust
    let new_ids: Vec<i64> = diesel::insert_into(crates)
        .values(versions)
        .on_conflict_do_nothing()
        .returning(id)
        .get_results(conn)
        .unwrap();
```

...then it does not compile:

```
   --> src/main.rs:107:10                                                                                                           
    |                                                                                                                               
107 |         .on_conflict_do_nothing()                                                                                             
    |          ^^^^^^^^^^^^^^^^^^^^^^                                                                                               
    |                                                                                                                               
    = note: the method `on_conflict_do_nothing` exists but the following trait bounds were not satisfied:                           
            `diesel::insertable::OwnedBatchInsert<diesel::query_builder::ValuesClause<(diesel::insertable::ColumnInsertValue<schema::crates::columns::name, diesel::expression::bound::Bound<diesel::sql_types::Text, std::string::String>>, diesel::insertable::ColumnInsertValue<schema::crates::columns::version, diesel::expression::bound::Bound<diesel::sql_types::Text, std::string::String>>, diesel::insertable::ColumnInsertValue<schema::crates::columns::checksum, diesel::expression::bound::Bound<diesel::sql_types::Text, std::string::String>>, diesel::insertable::ColumnInsertValue<schema::crates::columns::yanked, diesel::expression::bound::Bound<diesel::sql_types::Bool, bool>>), schema::crates::table>> : diesel::query_builder::UndecoratedInsertRecord<schema::crates::table>`
```

However, the mysterious part is that if I remove the `on_conflict_do_nothing()` then it DOES compile with the vec passed by value:

```rust

    let new_ids: Vec<i64> = diesel::insert_into(crates)
        .values(versions)
        .returning(id)
        .get_results(conn)
        .unwrap();
```

So I don't know why this is inconsistent or what is making it inconsistent, but it's weird enough that it seems like it should be wrong.

Thank you!

**Comment by weiznich** [comment]
Seems like there is some impl missing.

In detail, the corresponding impl for [`OwnedBatchInsert`](https://github.com/diesel-rs/diesel/blob/master/diesel/src/insertable.rs#L294) is missing [here](https://github.com/diesel-rs/diesel/blob/master/diesel/src/query_builder/insert_statement/mod.rs#L408).

Fixing this should be not hard:
1. Change `OwnedBatchInsert` to be generic over the table, similar to [`BatchInsert`](https://github.com/diesel-rs/diesel/blob/master/diesel/src/insertable.rs#L267)
2. Fix all compiler errors about a missing type paramater for `OwnedBatchInsert`. In most cases the table parameter is already there
3. Add the missing impl for `UndecoratedInsertRecord` in src/query_builder/insert_statement/mod.rs
4. Add a test case to [diesel_tests](https://github.com/diesel-rs/diesel/blob/master/diesel_tests/tests/update.rs#L230)


As a workaround: Pass the values list by reference.

**Comment by weiznich** [comment]
Already merged and released as 1.4
